libsrc/osic1p/cgetc.s

```diff
@@ -0,0 +1,29 @@
+;
+; char cgetc (void);
+;
+        .export         _cgetc
+        .import         cursor
+
+        .include        "osic1p.inc"
+        .include        "extzp.inc"
+
+; Input routine from 65V PROM MONITOR, show cursor if enabled
+_cgetc:
+        lda     cursor          ; show cursor?
+        beq     nocursor
+        ldy     CURS_X
+        lda     (SCREEN_PTR),y  ; fetch current character
+        sta     CURS_SAV        ; save it
+        lda     #$A1            ; full white square
+        sta     (SCREEN_PTR),y  ; store at cursor position
+nocursor:
+        jsr     INPUTC
+        pha                     ; save retrieved character
+        lda     cursor          ; was cursor on?
+        beq     nocursor2
+        lda     CURS_SAV        ; fetch saved character
+        ldy     CURS_X
+        sta     (SCREEN_PTR),y  ; store at cursor position
+nocursor2:
+        pla                     ; restore retrieved character
+        rts

```

Comment:
- Setting the Z flag for the `beq nocursor2`can be done without destroying A. This way the cycles to restore it need only be spent when the cursor actually needs to be removed.
- Functions declared to return an (unsigned) char still need to return a valid (unsigned) int !!! Therefore you'll see quite some `ldx #$00`in function epilogs.
- If you use X to check for `cursor` then you can avoid the cycles for the `ldx #$00` in case `cursor`is false.
```
nocursor:
  jsr INPUTC
  ldx cursor
  beq done
  tax
  lda tmp1
  ldy CURS_X
  sta (SCREEN_PTR),y
  txa
  ldx #$00
done:
  rts
```

You may think this is "over optimizing" but that's in general the amount of optimization applied to cc65 C library code.